### PR TITLE
Splitting offset multiplier nodes

### PR DESCRIPTION
**Pull request by VMatthijs**
This PR addresses issue https://github.com/stan-dev/stanc3/issues/16 .

It has the following side effects:
- slightly more precise error messages for bounds and offset-multiplier;
- made the `originblock` type a local data structure to the semantic check, rather than exposing it in the AST;
- slight refactoring and more code reuse in semantic check.

### CODE REVIEW

File: lib/Ast.ml [hunk 1]
```diff
@@ -11,19 +11,8 @@ type location =
 (** Delimited locations *)
 type location_span = {begin_loc: location; end_loc: location}
 
-(** Origin blocks, to keep track of where variables are declared *)
```
Comment (bob-carpenter) [inline lib/Ast.ml]: There's lingering doc on line 94 that refers to origin blocks, but there are no more origin blocks.  This is, by the way, what I meant about it being hard to keep the doc up to date in these cases and why I prefer not to have doc that repeats what the code's doing.

File: lib/Pretty_printing.ml [hunk 2]
```diff
@@ -171,6 +161,8 @@ and pretty_print_transformation = function
   | LowerUpper (e1, e2) ->
       "<lower=" ^ pretty_print_expression e1 ^ ", upper="
       ^ pretty_print_expression e2 ^ ">"
+  | Offset e -> "<offset=" ^ pretty_print_expression e ^ ">"
```
Comment (bob-carpenter) [inline lib/Pretty_printing.ml]: [question]
Is this going to get refactored to use the `Format` library?  I'd think our pretty printing is more important for readability than our C++ code generation (for things like line breaking).

File: lib/Pretty_printing.ml [hunk 3]
```diff
@@ -213,7 +205,8 @@ and pretty_print_transformed_type st trans =
   in
   match trans with
   | Identity -> pretty_print_sizedtype st
-  | Lower _ | Upper _ | LowerUpper _ | OffsetMultiplier _ ->
+  | Lower _ | Upper _ | LowerUpper _ | Offset _ | Multiplier _
+   |OffsetMultiplier _ ->
```
Comment (bob-carpenter) [inline lib/Pretty_printing.ml]: Space-vbar should be transposed so it lines up and is `| OffsetMultiplier`

File: lib/Semantic_check.ml [hunk 4]
```diff
@@ -19,6 +19,17 @@ open Pretty_printing
    on (1) a global symbol table vm, and (2) structure of type context_flags_record
    to communicate information down the AST. *)
 
+(** Origin blocks, to keep track of where variables are declared *)
+type originblock =
```
Comment (bob-carpenter) [inline lib/Semantic_check.ml]: I see `originblock` moved to here, but the thing with doc in the other file commented above didn't seem to have an origin block in it.

File: lib/Ast.ml (hunk 1 above)
Comment (VMatthijs) [reply lib/Ast.ml]: Right. That makes sense. I'll remove it.

File: lib/Pretty_printing.ml (hunk 2 above)
Comment (VMatthijs) [reply lib/Pretty_printing.ml]: Some day, if anyone is brave enough / feels like it. I agree it should be refactored.

File: lib/Pretty_printing.ml (hunk 3 above)
Comment (VMatthijs) [reply lib/Pretty_printing.ml]: Sadly, this is what OCamlformat makes of it.

File: lib/Semantic_check.ml (hunk 4 above)
Comment (VMatthijs) [reply lib/Semantic_check.ml]: Sorry. That comment was outdated.

**Comment by VMatthijs** [comment]
Thanks for reviewing, @bob-carpenter !

**Comment by bob-carpenter** [comment]
Should we file a bug report with OCamlformat?  This can't be their intended output, can it?  

Either way, no big deal for this PR if the autoformatter takes it out of our control.

**Comment by seantalts** [comment]
I think that's intentional from the formatter - the weirdly indented line is actually a continuation of the line above it, which doesn't have an action specified yet.

**Comment by VMatthijs** [comment]
Could someone maybe reapprove so this can be merged?

**Comment by VMatthijs** [comment]
Thanks!

**Pull request merged**